core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala

```diff
@@ -205,11 +205,45 @@ class KryoSerializationStream(
 
   private[this] var kryo: Kryo = serInstance.borrowKryo()
 
+  // This is only used when we write object and class separately.
+  var classWrote = false
```

Comment:
I don't see why you need this state and need to repeat the logic about writing / not writing classes everywhere. Surely this just goes in one writeAll / asIterator pair?

Reply:
Yeah, it used for `writeAll / asIterator`.  But for `MemoryStorea.putIteratorAsBytes`, we don't use the writeAll, we use this state to indicate whether we have written the class first.

Reply:
Why not write that state as an iterator of stuff, if that's how it behaves? rather than duplicate code. 'values' is already an iterator there.